### PR TITLE
TINY-6207: Fix font select menu items not handling multiple fonts in selection correctly

### PR DESCRIPTION
**Pull request by ltrouton**
Description of Changes:
* Updated the font family and font size detection code to check if multiple font properties are present in a selection. If so, then the menus will be blank instead of just showing one of the present font properties.

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] ~~Branch prefixed with `feature/` for new features (if applicable)~~
* [X] ~~License headers added on new files (if applicable)~~

Review:
* [X] Milestone set
* [ ] Review comments resolved

### CODE REVIEW

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 1]
```diff
@@ -44,18 +72,33 @@ const normalizeFontFamily = (fontFamily: string) =>
...
-const getFontProp = (propName: string) => (rootElm: Element, elm: Node): string => Optional.from(elm)
-  .map(SugarElement.fromDom)
-  .filter(SugarNode.isElement)
-  .bind((element: any) => getSpecifiedFontProp(propName, rootElm, element.dom)
-    .or(getComputedFontProp(propName, element.dom)))
-  .getOr('');
+const getFontInfo = (propName: FontProp, normalize: (str: string) => string) => (rootNode: Element, rngOrNode: Range | Node) => {
+  const fontProps = Arr.map(collect(DOMUtils.DOM, rngOrNode, rootNode, propName), normalize);
+  // If there is more than one font prop found, cannot declare that there is a single font
+  if (fontProps.length === 0 || fontProps.length > 1) {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:90]: This could be flipped and simplified to a ternary. eg:
```
return fontProps.length === 1 ? fontProps[0] : '';
```

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 2]
```diff
@@ -44,18 +72,33 @@ const normalizeFontFamily = (fontFamily: string) =>
...
+  Optional.from(DOMUtils.DOM.getStyle(elm, propName, true));
 
-const getFontProp = (propName: string) => (rootElm: Element, elm: Node): string => Optional.from(elm)
-  .map(SugarElement.fromDom)
-  .filter(SugarNode.isElement)
-  .bind((element: any) => getSpecifiedFontProp(propName, rootElm, element.dom)
-    .or(getComputedFontProp(propName, element.dom)))
-  .getOr('');
+const getFontInfo = (propName: FontProp, normalize: (str: string) => string) => (rootNode: Element, rngOrNode: Range | Node) => {
+  const fontProps = Arr.map(collect(DOMUtils.DOM, rngOrNode, rootNode, propName), normalize);
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:88]: Why are we using the top-level page DOM utils here? That is almost definitely a bug.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 3]
```diff
@@ -44,18 +72,33 @@ const normalizeFontFamily = (fontFamily: string) =>
...
+  // If there is more than one font prop found, cannot declare that there is a single font
+  if (fontProps.length === 0 || fontProps.length > 1) {
+    return '';
+  } else {
+    return fontProps[0];
+  }
+};
 
-const getFontSize = getFontProp('font-size');
+const getFontSize =
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:97]: This is a pretty weird code style. Why was this split over 2 lines? I could understand if it was the internals of a function block, but not for an assignment.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 4]
```diff
@@ -5,26 +5,54 @@
...
+      const matches = dom.getParents(node, '*[style],font', limit);
+      Arr.findMap(matches, (match) => getSpecifiedFontProp(propName, SugarElement.fromDom(match)))
+        .orThunk(() => {
+          // Cannot find computed style on text node so have get parent node if required
+          const elm = NodeType.isText(node) ? dom.getParent(node, '*', rootNode) : node as Element;
+          return getComputed ? getComputedFontProp(propName, elm) : Optional.none();
+        })
+        .each((value) => {
+          // Make sure value is unique
+          if (Arr.forall(collector, (existing) => existing !== value)) {
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:36]: We can't use `Set` it's not available on IE 11. You could use an object for this though instead.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 5]
```diff
@@ -5,26 +5,54 @@
...
-      return Obj.get(legacyPropNames, propName).bind((legacyPropName) => Attribute.getOpt(elm, legacyPropName));
-    } else {
-      return Optional.none();
-    }
-  });
-  const isRoot = (elm: SugarElement) => Compare.eq(SugarElement.fromDom(rootElm), elm);
+const collect = (dom: DOMUtils, rngOrNode: Range | Node, rootNode: Element, propName: FontProp): string[] => {
+  const collector: string[] = [];
+
+  const process = (nodes: Node[], limit: Node, getComputed: boolean) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:25]: `limit` is a rather strange name for a node, I'd suggest calling this `root` instead, as that's what we use in almost every other place.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:36]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set apparently it is, there are just a few caveats (but easy to work around)

File: modules/tinymce/src/core/main/ts/commands/FontCommands.ts [hunk 6]
```diff
@@ -58,23 +45,33 @@ const normalizeFontNames = (font: string) => {
   }).join(',');
 };
 
+const fontQuery = (editor: Editor, fontProp: FontInfo.FontProp) => {
+  const getter = fontProp === 'font-family' ? FontInfo.getFontFamily : FontInfo.getFontSize;
+  const selection = editor.selection.getRng().cloneRange();
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/commands/FontCommands.ts:50]: This name is rather confusing, as this isn't a `Selection` object, it's a `Range` object. I'd suggest calling this `rng` or `range`.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts (hunk 5 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:25]: Also from what I can see, you don't need a `Node[]` here and instead just require an `ArrayLike<Node>` array instead. That'll save having to do the `Arr.from(node.childNodes)` below as well. I could have very well missed something though :shrug:

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 7]
```diff
@@ -5,26 +5,54 @@
...
+    Arr.each(nodes, (node) => {
+      const matches = dom.getParents(node, '*[style],font', limit);
+      Arr.findMap(matches, (match) => getSpecifiedFontProp(propName, SugarElement.fromDom(match)))
+        .orThunk(() => {
+          // Cannot find computed style on text node so have get parent node if required
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:30]: There's a typo in this comment by the looks.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts [hunk 8]
```diff
@@ -5,26 +5,54 @@
...
+          };
+        });
+
+      // Don't need to worry about potentially getting the computed value for children as if no fonts are found on the child,
+      // it will inherit the 'node' font value which may be computed
+      process(Arr.from(node.childNodes), node, false);
+    });
+  };
+
+  if (isNode(rngOrNode)) {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:47]: Hmm... this will cause very different behaviours, as `RangeWalk` walks from the bottom up. It's also going to cause it to walk from the bottom, up one level and then walk back down since `process` recursively checks child nodes.

That doesn't seem like a very good solution for collecting the fonts here. I'll have to think of other options, but the first option that comes to mind is when a node is passed, create a range around that node and remove the recursive calls in `process`.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/fmt/FontInfo.ts:36]: Right, I just remembered we'd avoided it because it was an issue on IE 11.

File: modules/tinymce/src/core/test/ts/browser/FontSelectTest.ts [hunk 9]
```diff
@@ -14,80 +30,157 @@ UnitTest.asynctest('browser.tinymce.core.FontSelectTest', function (success, fai
...
       Chain.mapper(Fun.compose(Strings.trim, TextContent.get)),
       Assertions.cAssertEq('Should be the expected display value', expectedValue)
     ]);
-  };
 
-  TinyLoader.setupLight(function (editor, onSuccess, onFailure) {
+  TinyLoader.setupLight((editor: Editor, onSuccess, onFailure) => {
     const tinyApis = TinyApis(editor);
 
+    const sAssertFontsForSelections = (selections: Sel[], expectedFontSize: string, expectedFontFamily: string): Step<unknown, unknown>[] =>
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/test/ts/browser/FontSelectTest.ts:43]: Yeah, I'd prefer to see this wrapped int a `GeneralSteps.sequence` as well. It's less to add else where and makes it a little more readable as you don't need to use a spread operator everywhere.

File: modules/tinymce/src/core/test/ts/browser/fmt/FontInfoTest.ts [hunk 10]
```diff
@@ -138,12 +139,14 @@ UnitTest.asynctest('browser.tinymce.core.fmt.FontInfoTest', function (success, f
     iframe.contentDocument.close();
   });
 
-  suite.test('comments should always return empty string', function () {
-    assertComputedFontProp('fontFamily', '<!-- comment -->', [ 0 ], '');
-    assertComputedFontProp('fontSize', '<!-- comment -->', [ 0 ], '');
-    assertSpecificFontProp('fontFamily', '<!-- comment -->', [ 0 ], '');
-    assertSpecificFontProp('fontSize', '<!-- comment -->', [ 0 ], '');
-  });
+  if (!Env.browser.isIE()) {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/fmt/FontInfoTest.ts:142]: Why was this added? Does this mean it doesn't work on IE/returns different results? This smells like a potential regression/bug.

File: modules/tinymce/src/core/main/ts/commands/FontCommands.ts [hunk 11]
```diff
@@ -58,23 +45,33 @@ const normalizeFontNames = (font: string) => {
...
+  const isTextSelection = NodeType.isText(selection.startContainer) && selection.startContainer === selection.endContainer;
+  // Have to directly use text node instead of selectionRng for text selection as there are cases where
+  // RangeWalker used in the getter fn will not return results otherwise
+  const rngOrNode: Range | Node = isTextSelection ? selection.startContainer : selection;
+  if (rngOrNode) {
+    return getter(editor.getBody(), rngOrNode);
+  } else {
+    return '';
```
[the remark this answers is not in the document]
Comment (ltrouton) [reply modules/tinymce/src/core/main/ts/commands/FontCommands.ts]: I had forgotten to remove the check after removing some alternate code I was working on. Thanks for picking this up.

File: modules/tinymce/src/core/main/ts/fmt/FontInfo.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ltrouton) [reply modules/tinymce/src/core/main/ts/fmt/FontInfo.ts]: Using `Set` did cross my mind but I noticed we didn't use it anywhere in the code. Anyways, have swapped out using an array for an object as the collector now.

File: modules/tinymce/src/core/test/ts/browser/FontSelectTest.ts [hunk 12]
```diff
@@ -14,80 +30,157 @@ UnitTest.asynctest('browser.tinymce.core.FontSelectTest', function (success, fai
...
+        tinyApis.sNodeChanged(),
+        sAssertSelectBoxDisplayValue('Font sizes', expectedFontSize),
+        sAssertSelectBoxDisplayValue('Fonts', expectedFontFamily)
+      ]);
+
     Pipeline.async({}, [
-      Logger.t('Font family and font size on initial page load', GeneralSteps.sequence([
-        sAssertSelectBoxDisplayValue(editor, 'Font sizes', '12px'),
-        sAssertSelectBoxDisplayValue(editor, 'Fonts', 'Arial')
+      Log.step('', 'Font family and font size on initial page load', GeneralSteps.sequence([
```
[the remark this answers is not in the document]
Comment (ltrouton) [reply modules/tinymce/src/core/test/ts/browser/FontSelectTest.ts]: I should have taken more notice of the existing code to see it could be simplified further. Thanks for pointing that out.

**Comment by ltrouton** [comment]
Closing this as we have discussed as team and think we need to come up with a more performant and generic solution for detecting styles of interest over the whole selection. More details are on the related Jira ticket.
